### PR TITLE
fix(security): complete #206 hardening bundle

### PR DESCRIPTION
**Pull request by osirison**
Completes the low-severity hardening bundle from the 2026-07-02 security review (#206). Two items were already on the branch (client-facing errors, `/health/metrics` token); this finishes the remaining ones and hardens the base.

## Items

- **Client-facing error hygiene** — MCP tool handlers now return a generic message and keep Prisma/Redis/vector internals server-side (`security/client-error.util.ts`); Zod validation + authored `ClientFacingError` messages still surface.
- **`/health/metrics` scrape token** — optional `MetricsTokenGuard`: open when `METRICS_TOKEN` is unset, `401` otherwise (`Authorization: Bearer` or `X-Metrics-Token`, constant-time).
- **Helmet CSP** — replaced `contentSecurityPolicy: false` with a restrictive `default-src 'none'` policy fit for a JSON/MCP API (`security/security-headers.util.ts`).
- **`recall` userId latent trap** — documented the Zod-version dependency in `schemaAcceptsUserId` and added a guard test using a `.refine()`-wrapped identity schema, so a Zod downgrade / effects-wrapped tool that silently disables userId injection fails CI instead of leaking cross-tenant.
- **GitHub OAuth email (default-deny)** — `isGithubEmailVerified` resolves `/user/emails` and requires the address present with `verified: true`; fail-closed on any error. GitHub is already a hard login dependency, so no new availability surface.
- **Prompt-injection delimiting** — `prompt_context` / `compress_context` / `load_context` / `reflect` wrap stored content in an untrusted-content fence + notice, neutralise forged markers, and reserve the envelope from the char/token budgets (the `estimatedTokens ≤ budget` invariant still holds).
- **Read-only web DB role** — `docker/postgres/10-web-readonly-role.sh` provisions an `engram_readonly` role (env-driven password, `format(%I,%L)`, idempotent `\gexec`); the dashboard prefers `WEB_DATABASE_URL` and falls back to `DATABASE_URL`. Validated live: role can `SELECT`, denied `INSERT`.
- **`AUTH_URL` guidance** — documented in `.env.example` for prod behind a proxy (`trustHost: true`).

## ⚠️ Breaking change (prod compose)

`docker-compose.prod.yml` now **requires** `WEB_DB_READONLY_PASSWORD` (`:?`). Existing deployments must set it before `docker compose up`. It is consumed only on a fresh DB volume (initdb), but the check is at container start — set it to a strong value and reuse it in `WEB_DATABASE_URL`.

## Verification

`pnpm build` (packages) · `pnpm lint` (17/17) · `pnpm typecheck` · `pnpm docs:check` — all green. Tests: mcp-server 519, `@engram/core` 36 (incl. the recall guard), web 66 (incl. 11 oauth-verify). The read-only role's SELECT-yes / INSERT-denied behaviour was verified against the running Postgres.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: apps/mcp-server/src/memory/memory.service.ts [hunk 1]
```diff
@@ -1028,7 +1065,13 @@ export class MemoryService {
...
+    // Frame the entries as untrusted content (#206). Reserve the envelope's
+    // length from maxChars so the assembled block still respects the budget.
+    const openText = `## Memory Context\n${UNTRUSTED_CONTENT_NOTICE}\n\n${UNTRUSTED_CONTENT_OPEN}\n`;
+    const closeText = `${UNTRUSTED_CONTENT_CLOSE}\n`;
+    const entriesBudget = maxChars - openText.length - closeText.length;
+
+    const lines: string[] = [];
     let used = 0;
     let included = 0;
     let truncated = false;
```
[the remark this answers is not in the document]
Comment (osirison) [reply apps/mcp-server/src/memory/memory.service.ts:1080]: Good catch — fixed in c847404.

The untrusted-content framing (notice + fence markers) is a fixed, mandatory ~210-char cost, so a `maxChars` below it genuinely cannot be honored. Rather than let the block silently exceed the budget, I raised the `maxChars` floor for `compress_context`/`load_context` from 100 to **512** (comfortably above the envelope), so `entriesBudget` is always positive and the assembled block never exceeds `maxChars`. I also switched the entry join from `\n` to `` (entries already end in `\n`) so `charCount = openText + used + closeText ≤ maxChars` exactly — no separator overage.

Added tests: a min-budget boundary case (`maxChars: 512` → framed + `charCount ≤ 512`) and a DTO test asserting the schema rejects `maxChars < 512`. The token-based `prompt_context` was already safe (its ~52-token envelope fits under the min-100 *token* budget).
